### PR TITLE
Basic built-in image I/O system #507

### PR DESCRIPTION
**Pull request by archibate**
Related issue id = #507

### CODE REVIEW

**Comment by archibate** [comment]
```python
# image_display.py
import taichi as ti
import numpy as np
import sys

ti.init(arch=ti.x86_64)

filename = sys.argv[1]
x = ti.imread(filename, channels=3)
x = x.astype(np.float32) / 255

gui = ti.GUI(filename, res=x.shape[:2])
while not gui.has_key_event() or gui.get_key_event().key == ti.GUI.MOTION:
  gui.set_image(x)
  gui.show()
```

File: python/taichi/misc/gui.py [hunk 1]
```diff
@@ -53,7 +53,10 @@ def set_image(self, img):
       img = img.to_numpy()
     assert isinstance(img, np.ndarray)
     assert len(img.shape) in [2, 3]
-    img = img.astype(np.float32)
+    if img.dtype not in [np.float32, np.float64]:
+      img = img.astype(np.float32) * (1.0 / 255)
+    else:
+      img = img.astype(np.float32)
```
Comment (archibate) [inline python/taichi/misc/gui.py]: Do you think this should be here, or in `cook_image`? Or add `set_image_raw(u8 *img)` in pybind11?

Comment (yuanming-hu) [reply python/taichi/misc/gui.py]: You mean we should invoke `cook_image` in `GUI.set_image`? I think that's a great idea!

Comment (archibate) [reply python/taichi/misc/gui.py]: Yeah it is. But I mean, is `GUI.set_image` expected to support `np.uint8` format image?

Comment (yuanming-hu) [reply python/taichi/misc/gui.py]: Yes :-)

**Comment by yuanming-hu** [comment]
Thanks for the PR! This will be an extremely helpful feature to Taichi users. Could you add a basic test case? Basically
 - Generate a random 128x128 image 
 - Write the image to disk using your `imwrite`
 - Read the image from disk using your `imread`
 - Test if the original image is the same as the loaded one

Please consider different data types (u8, u16, u32, u64, f32, f64) and image formats `png, bmp, jpg`. Note that `jpg` is lossy, and there might be subtle differences introduced by compression.

**Comment by archibate** [comment]
When trying to use `ti.random()` for `u8`:
```
Traceback (most recent call last):
  File "image_io.py", line 11, in <module>
    img = pixel.to_numpy()
  File "/home/bate/Develop/taichi/python/taichi/lang/expr.py", line 316, in to_numpy
    tensor_to_ext_arr(self, arr)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 347, in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "/home/bate/Develop/taichi/python/taichi/lang/kernel.py", line 205, in materialize
    visitor.visit(tree)
  File "/usr/lib/python3.8/ast.py", line 360, in visit
    return visitor(node)
  File "/home/bate/Develop/taichi/python/taichi/lang/transformer.py", line 449, in visit_Module
    self.generic_visit(node)
  File "/home/bate/Develop/taichi/python/taichi/lang/transformer.py", line 78, in generic_visit
    value = self.visit(value)
  File "/usr/lib/python3.8/ast.py", line 360, in visit
    return visitor(node)
  File "/home/bate/Develop/taichi/python/taichi/lang/transformer.py", line 497, in visit_FunctionDef
    assert False
AssertionError
```
Something not done in #505 (stage 1), which also don't support casting for smaller types.

**Comment by archibate** [comment]
Finished! For `jpg`, instead of `assert pixel == pixel_r`, I just `imshow` them in two window and let the user to compare them by eyes!

File: python/taichi/lang/transformer.py [hunk 2]
```diff
@@ -485,16 +485,19 @@ def visit_FunctionDef(self, node):
...
+              ti.i8: 'ti.i8',
+              ti.i16: 'ti.i16',
+              ti.i32: 'ti.i32',
+              ti.i64: 'ti.i64',
+              ti.u8: 'ti.u8',
+              ti.u16: 'ti.u16',
+              ti.u32: 'ti.u32',
+              ti.u64: 'ti.u64',
+          }
+          dt = self.parse_expr(dt_table[array_dt])
```
Comment (archibate) [inline python/taichi/lang/transformer.py]: You may want more like this for adding new data types!

Comment (archibate) [inline python/taichi/lang/transformer.py]: You may want more like this for adding new data types!

Comment (yuanming-hu) [reply python/taichi/lang/transformer.py]: Cool!

Comment (yuanming-hu) [reply python/taichi/lang/transformer.py]: We can also make use of `ti.core.data_type_short_name`.

**Pull request merged**